feat : 집계 기능 - 멀티스레드 버전

gutenLEE opened the pull request:
## 개요
- 관련 Issue 번호 : #82 

## 작업사항
- 멀티 스레드로 집계 기능 추가
- 전략 패턴으로 쿼리 집계, 멀티스레드 집계 모듈 동적으로 변경하도록 설계

gutenLEE committed:
feat : 집계 기능 - 멀티스레드 버전

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/MultithreadingAggregationService.java:
```diff
@@ -0,0 +1,54 @@
+
+    }
+
+    private MultithreadingStrategy getStrategy(boolean multithreading) {
+        if (multithreading){
+            return new FirstMultithreadingStrategy(aggregationMapper);
+        }
+        return new NonMultithreadingStrategy(aggregationMapper);
+    }
+}
```
전략패턴으로 집계 로직 동기 처리, 멀티스레딩 버전 두 가지 구현체 만들었습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/FirstMultithreadingStrategy.java:
```diff
@@ -0,0 +1,42 @@
+public class FirstMultithreadingStrategy implements MultithreadingStrategy {
+
+    private final TicketOrderAggregationMapper aggregationMapper;
+    @Override
+    public void execute(ExecutorService es, List<Ticket> tickets) {
+
+        Map<String, List<Ticket>> collected
+                = tickets.stream()
+                    .collect(groupingBy(Ticket::getGradeName));
+
+        collected.entrySet()
+                .parallelStream()
+                .forEach(entry -> {
+                    List<Long> seqList = entry.getValue().stream().map(Ticket::getSeq).collect(Collectors.toList());
+                    List<AggregateSelectMapperDto.Response> responses = aggregationMapper.selectByTicketSeqList(seqList);
+                    log.info("order Try {}", responses.size());
+                });
+    }
```
결과 값을 어떻게 리턴할 지 결정하지 못해 void로 남겨두었습니다. 
좀 더 학습후에 개선해보겠습니다.

f-lab-maverick replied:
1. 멀티스레드로 집계를 한다면 좀 더 유의미한 작업의 단위로 나뉘어야 할 것으로 보입니다. 작업 단위가 커보이네요.
2. 현재 단순히 stream을 병렬로 처리하고 있는데요, 이럴 경우 기본적으로 공유 fork join pool을 사용하기 때문에 별도 스레드풀에서 돌려야 다른 작업의 영향을 안받을 수 있습니다.
3. 또한 현재 파라미터로 받은 `es`를 사용하지 않는 것으로 보이네요. custom thread pool로도 패러렐 스트림을 돌릴 수 있으니 찾아보셔도 되고, 작업을 직접 스레드풀에 할당해주셔도 됩니다.

gutenLEE commented on src/test/java/com/heeverse/ticket_order/service/reader/SimpleAggregationReaderTest.java:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    @Test
+    @DisplayName("비정규화 테이블에서 처리")
+    void queryAggrTest() throws Exception {
+        // DB CPU 38%, TPS 6/sec
+        for (int i = 0; i < LOOP; i++) {
+            List<AggregateSelectMapperDto.Response> deNormalization
+                    = aggregationMapper.selectGroupByGradeNameDeNormalization(1L);
+            log.info("{}", i);
+        }
+    }
+
+    @Test
+    @DisplayName("정규화 테이블에서 처리")
+    void queryAggrNormalizationTest() throws Exception {
+        //  CPU 86%, TPS 1.3/sec
+        for (int i = 0; i < LOOP; i++) {
+            List<AggregateSelectMapperDto.Response> normalization
+                    = aggregationMapper.selectGroupByGradeName(1L);
+            log.info("{}", i);
+        }
+    }
```
정규화된 테이블과 역정규화된 테이블에서 집계 처리 시 
CPU나 TPS나 역정규화된 테이블에서 집계 처리가 성능상 이점을 보였습니다.

gutenLEE commented on src/test/java/com/heeverse/ticket_order/service/reader/SimpleAggregationReaderTest.java:
```diff
@@ -0,0 +1,106 @@
+
+@ActiveProfiles("dev-test")
+@SpringBootTest
+@EnabledOnOs({OS.MAC})
```
학습 테스트라서 테스트 서버에서는 돌지 않도록 하기 위함입니다.

gutenLEE commented on src/main/resources/mapper/TicketOrderAggregationtMapper.xml:
```diff
@@ -53,7 +53,20 @@
+    <!--티켓 등급별 예매 시도 조회-->
+    <select id="selectByTicketSeqList" parameterType="java.lang.Long"
+            resultType="com.heeverse.ticket_order.domain.dto.persistence.AggregateSelectMapperDto$SimpleResponse">
+        SELECT
+            t.ticket_seq
+        FROM
+            ticket_order_log t
+        WHERE ticket_seq IN
+            <foreach collection="list" item="ticketSeq" separator="," open="(" close=")">
+                #{ticketSeq}
+            </foreach>
```
멀티스레딩 집계 버전에서 사용할 쿼리를 고민중입니다. 
LEFT JOIN을 아예 없앨지, LEFT JOIN을 하자만 GROUP BY를 어플리케이션에서 처리할지
둘 다 비교해보겠습니다.

gutenLEE reviewed (commented):
현재 사용한 예매 로그 테이블 레코드는 20만개였습니다. 
고척돔 공연장 16,000 좌석에서 좌석 당 100번의 예매 시도가 발생해도 총 1,600,000개 레코드가 생성됩니다. 
최소 1,000,000개 레코드로 집계 처리를 했을때 유의미한 결과가 나올 것 같습니다. 
현재 진행중인 테스트가 완료되면 진행해보겠습니다.

gutenLEE commented on src/test/java/com/heeverse/ticket_order/service/reader/SimpleAggregationReaderTest.java:
```diff
@@ -0,0 +1,106 @@
+    private static final int THREAD_POOL_SIZE = 100;
+
+
+    @Test
+    @DisplayName("멀티스레드로 집계")
+    void multithreadingAggregationTest() throws Exception {
+
+        //  1 min 40 sec, DB CPU 33%
+        // 1 min 33 sec,  DB CPU 33%
+        reader.setStrategy(new FirstMultithreadingStrategy(aggregationMapper));
+
+        for (int i = 0; i < LOOP; i++) {
+            List<AggregateSelectMapperDto.Response> responseList
+                    = reader.getResultGroupByGrade(new AggregateSelectMapperDto.Request(1L));
+            log.info( "===== {} ====", i);
+        }
+    }
+
+    @Test
+    @DisplayName("동기적으로 집계 처리")
+    void synchronousAggregationTest() throws Exception {
+
+        reader.setStrategy(new NonMultithreadingStrategy(aggregationMapper));
+        // DB CPU 25%
+        for (int i = 0; i < LOOP; i++) {
+            List<AggregateSelectMapperDto.Response> responseList
+                    = reader.getResultGroupByGrade(new AggregateSelectMapperDto.Request(1L));
+            log.info( "===== {} ====", i);
+        }
+    }
```
동기적으로 코드 수행이 더 빠릅니다 현재.
관련 도서를 참고해보니 I/O 처럼 순차 처리 비율이 얼마나 되는지 파악이 필요함을 알았습니다.

f-lab-maverick commented on src/main/java/com/heeverse/common/ExecutorUtils.java:
```diff
@@ -0,0 +1,20 @@
+@UtilityClass
+public class ExecutorUtils {
+
+    public static ExecutorService createFixedThreadPool(final int threads) {
+        return Executors.newFixedThreadPool(threads);
+    }
```
이 메서드는 `Executors.newFixedThreadPool`을 래핑한 클래스인 것으로 보이는데요, 굳이 이렇게 래핑하지 않아도 `Executors`를 동일하게 사용할 수 있을 것으로 보입니다. 이렇게 해둔 이유가 있을까요?

gutenLEE replied:
전략객체로 여러 버전의 집계 구현체를 개발해 둔 상태라 일관된 방식으로 `Executors` 호출하기 위해서 
메서드로 빼봤습니다. 리팩토링하며 피드백 주신 방향도 고민해보겠습니다

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/SimpleAggregationReader.java:
```diff
@@ -0,0 +1,44 @@
+
+    public SimpleAggregationReader(TicketMapper ticketMapper) {
+        this.ticketMapper = ticketMapper;
+        this.es = createFixedThreadPool(3);
```
`ExecutorService`를 `Bean`으로 등록해 둔다면 생성자 주입을 받을 수 있습니다. 아니면 `Reader`별로 별도의 Tread pool을 사용할 의도로 이렇게 만들어두신 건가요?

gutenLEE replied:
빈으로 등록하는 방식을 사용했다가 Reader별로 스레드풀을 다르게 적용할 수도 있을듯하여 그랬습니다!

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/SimpleAggregationReader.java:
```diff
@@ -0,0 +1,44 @@
+        this.es = createFixedThreadPool(3);
+    }
+
+    public void setStrategy(MultithreadingStrategy strategy) {
+        this.strategy = strategy;
+    }
```
상당히 위험한 코드네요.
이 `SimpleAggregationReader`는 Bean으로 등록되어있기 때문에 Spring container 내부에서 싱글톤 패턴으로 공유되는 객체입니다. 그런 공유 객체에 setter를 열어두었다는 것은 내가 동작을 하고있던 도중이라도 언제든지 전략이 바뀔 수 있다는 뜻이에요.

전략에 따라 다르게 동작하는걸 원하신다면 `SimpleAggregationReader`를 Bean으로 공용으로 사용하시면 안됩니다.

gutenLEE replied:
아하 전략이 다이나믹하게 변경이 필요할 수도 있지 않을까 싶었는데
장점보다 사이드이펙트만 증가하겠네요..수정하겠습니다

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/FirstMultithreadingStrategy.java:
```diff
@@ -0,0 +1,42 @@
+                .parallelStream()
+                .forEach(entry -> {
+                    List<Long> seqList = entry.getValue().stream().map(Ticket::getSeq).collect(Collectors.toList());
+                    List<AggregateSelectMapperDto.Response> responses = aggregationMapper.selectByTicketSeqList(seqList);
```
MySQL의 In절에는 개수 제한이 있습니다. 이 개수제한이 몇 개인지를 확인해보시고, in 절 개수가 많을 경우 여러번에 걸쳐 나눠 쿼리하는 로직이 필요해보이네요~

f-lab-maverick reviewed (changes requested):
현재 구조에서는 병렬처리를 한다고 해서 성능이 뛰어나게 좋아지지 않을 것으로 보입니다.
fork-join이 어떻게 작업을 나누는지, 어떻게 빠른 속도를 만들 수 있는지 찾아보시고
병렬처리시 주의할 점들과 stream을 병렬로 돌릴 때 작업을 나누는 방법 등을 찾아보신 후 다시 적용해보시겠어요?

gutenLEE committed:
feat : 멀티스레딩 처리 초안 & 전략 객체 의존관계 주입 수정

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/StreamAggregationReader.java:
스트림 처리는 아직 학습중이라 코드만 옮겨놨습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
멀티스레드 처리 계획을 절차형으로 코드 작성해봤습니다. 

1. DB I/O 작업과 CPU 작업을 분리합니다.  (Pub-Sub 구조로 분리 생각중)
    -    DB조회 결과 Queue에 적재
3. DB 조회 결과를 멀티스레드로 집계합니다. 
    -  멀티스레드로 queue의 작업을 받아 집계 작업을 수행합니다. 
5. 집계 결과를 공유 객체에 저장합니다.
    멀티스레드의 동시성을 제어를 효과적으로 하는 Concurrent 패키지 클래스를 조사 후 적용할 계획입니다.

f-lab-maverick replied:
집계결과를 모으는 작업이 애매할 경우, DB에 저장하는 방법도 있습니다.
모든 멀티스레드 작업 결과를 모은 후 결과를 내야한다면 관련 컬랙션이 필요하기는 하겠네요~

gutenLEE commented on src/main/java/com/heeverse/common/ExecutorsComponent.java:
```diff
@@ -0,0 +1,29 @@
+        return Runtime.getRuntime().availableProcessors();
+    }
+
+    @Bean
+    public ExecutorService fixedPool() {
+        return Executors.newFixedThreadPool(availableCores() * 2);
+    }
```
병렬책을 읽으니 스레드 사이즈 설정은 순차처리 비율이나 다른 여러 조건도 고려해야 함을 배웠는데 
다양한 테스트후 감을 잡아야 할 것 같습니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/SingleThreadingStrategy.java:
```diff
@@ -0,0 +1,42 @@
+                    = aggregationMapper.selectByTicketSeqList(list.stream().map(Ticket::getSeq).collect(Collectors.toList()));
+            log.info("{} orderTry {}", entry.getKey(), responseList.size());
+        }
+    }
```
이 메서드는 아무것도 리턴하지 않는 것인가요?

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/MultithreadingAggregationService.java:
```diff
@@ -0,0 +1,55 @@
+            return new MultithreadingStrategy(aggregationMapper);
+        }
+        return new SingleThreadingStrategy(aggregationMapper);
+    }
```
현재 이 메서드가 존재하는 클래스 자체가 `MultithreadingAggregationService`로 멀티스레드를 기반으로 하는 집계 서비스입니다. 이 서비스 내부에서 멀티스레드로 돌릴지 말지 전략을 한번 더 선정해야할 이유가 있을까요?

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/MultiAggregationReader.java:
```diff
@@ -0,0 +1,40 @@
+        strategy.execute(es, tickets);
+        return null;
+    }
+}
```
리턴값이 컬랙션인 경우 차라리 비어있는 리스트를 내려주세요.
null을 내려주는 순간 NPE의 가능성이 굉장히 높아집니다.

사용하는 쪽에서는 List와 같은 컬랙션이 리턴되면 Not null이며, 최소 비어있는 리스트가 리턴될 것으로 기대하기 때문에 null은 지양해주세요.

리턴될 것이 없다면 비어있는 리스트라도 리턴해주시고, 만약 리턴할것이 정말 아무것도 없다면 메서드 시그니처를 수정하여 void 리턴타입으로 변경하는 것이 좋을 것으로 보입니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/SingleAggregationReader.java:
```diff
@@ -0,0 +1,40 @@
+        strategy.execute(es, tickets);
+        return null;
+    }
+}
```
싱글스레드 처리라면 굳이 비동기처리를 할 필요가 있나요?

gutenLEE committed:
feat : 집계 로직 객체관계 리스트럭처링

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/MultithreadingAggregationService.java:
```diff
@@ -22,19 +20,17 @@
-        // ticket
 
         try {
+
+            AggregationReader readerBean = getReaderBean(request.getStrategyType());
```
![MultithreadingStrategy](https://github.com/f-lab-edu/heeverse-ticket/assets/61870748/fbd8ed17-d04c-4e52-b1e5-37f2359f090e)

테스트 해보니 구조적으로 전략 객체를 변경하기 어려워 구조를 위 다이어그램으로 변경했습니다. 
RequestDto의 `StrategyType` 에 따라서 동적으로 `Reader` 구현체를 호출하게 변경했습니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/SingleAggregationReader.java:
```diff
@@ -0,0 +1,40 @@
+
+        List<Ticket> tickets = ticketMapper.findTickets(request.concertSeq());
+        strategy.execute(es, tickets);
+        return null;
```
여기서는 왜 null을 리턴하나요?

f-lab-maverick replied:
Return type이 컬랙션인데 null을 리턴하면 NPE 가능성이 상당히 높아집니다. empty list를 리턴하거나, 다른 방법을 고민해주세요~

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/StreamAggregationReader.java:
```diff
@@ -0,0 +1,47 @@
+                    List<AggregateSelectMapperDto.Response> responses = aggregationMapper.selectByTicketSeqList(seqList);
+                    log.info("{} : order Try {}", entry.getKey(), responses.size());
+                    return new AggregateSelectMapperDto.Response(request.concertSeq(), entry.getKey(), seqList.size(), responses.size());
+                }).collect(Collectors.toList());
```
stream 체이닝 망식 도중 코드블럭으로 로직이 나올 경우 가독성이 많이 떨어지는 상황이 발생하게 됩니다.
이럴경우 이런 로직을 메서드로 분리하여 stream 체이닝을 최대한 간결하게 유지해주는 것이 좋습니다.
```suggestion
                .map(~~method).collect(Collectors.toList());
```

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/StreamAggregationReader.java:
```diff
@@ -0,0 +1,47 @@
+        return collected.entrySet().parallelStream()
+                .map(entry -> {
+                    List<Long> seqList = entry.getValue().stream().map(Ticket::getSeq).collect(Collectors.toList());
+                    List<AggregateSelectMapperDto.Response> responses = aggregationMapper.selectByTicketSeqList(seqList);
```
MySQL In절 개수 제한은 확인해보셨나요?

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/StreamAggregationReader.java:
```diff
@@ -0,0 +1,47 @@
+                .collect(groupingBy(Ticket::getGradeName));
+
+
+        return collected.entrySet().parallelStream()
```
이렇게 병렬 스트림으로 돌릴 경우, 당연한 이야기겠지만 서로 다른 트랜잭션에서 동작하게 됩니다.
이럴 경우 동시성이슈나 타이밍 이슈는 없을까요?

gutenLEE replied:
멀티스레드 부분 작업 완료 후 보강하겠습니다!

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -0,0 +1,70 @@
+                result.putIfAbsent(poll.gradeName(), new AtomicLong(1));
+                result.computeIfPresent(poll.gradeName(), (a, b) -> {
+                    b.incrementAndGet();
+                    return b;
```
```suggestion
                    return b.incrementAndGet();
```

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -0,0 +1,70 @@
+
+    private final TicketOrderAggregationMapper aggregationMapper;
+    @Override
+    public List<AggregateSelectMapperDto.Response> execute(ExecutorService es, List<Ticket> tickets) {
```
이 메서드는 너무 커서 분리가 필요해보입니다.
메서드를 역할별로 분리해서 작은 단위로 만들어주세요

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -0,0 +1,70 @@
+                    return b;
+                });
+            });
+            if (queue.isEmpty())
+                latch.countDown();
```
이런 형식으로 단 하나의 락만 기다리는것이라면 latch를 사용할 이유가 없어보입니다.  latch를 사용한 이유가 어떤것일까요? 여러 동작들이 모두 완료되는것을 대기하는 용도로 보통 사용합니다만, 여기서는 비동기 동작을 대기하는 용도로만 사용하는것으로 보입니다. 그리고 그것을 queue의 size와 latch를 동시에 사용하고있어서 깔끔한 로직으로 보기 어렵네요.

이런 경우, 차라리 queue의 사이즈만큼 latch사이즈를 늘리고, `es.execute`로직 안에서 모든 로직 종료시 countDown을 해주는게 더 명시적이지 않을까요?

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/SingleThreadingStrategy.java:
```diff
@@ -0,0 +1,45 @@
+                    = aggregationMapper.selectByTicketSeqList(list.stream().map(Ticket::getSeq).collect(Collectors.toList()));
+            log.info("{} orderTry {}", entry.getKey(), responseList.size());
+        }
+        return null;
```
리턴타입으로 null은 사용하지 말아주세요~ 다른 리뷰에 관련 내용을 적어두었으니 확인 부탁드립니다.

gutenLEE replied:
넵!!

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/SingleThreadingStrategy.java:
```diff
@@ -0,0 +1,45 @@
+                .collect(groupingBy(Ticket::getGradeName));
+
+
+        ArrayList<AggregateSelectMapperDto.Response> result = new ArrayList<>();
```
이 변수는 사용하지 않는것인가요?

gutenLEE replied:
사용하는 코드인데 작업이 덜 되었네요 수정하겠습니다

f-lab-maverick reviewed (changes requested):
아직 로직이 완성되지 않은 부분들이 눈에 보이네요.
테스트 해보았을 때 성능차이가 어느정도 있었는지도 함께 첨부되면 좋을 것으로 보입니다~

gutenLEE committed:
feat : 페이징 추가 & 집계 결과 리턴 null 제거

gutenLEE commented on src/main/resources/mapper/TicketOrderAggregationtMapper.xml:
```diff
@@ -69,4 +70,21 @@
+              AND
+                seq > #{offset}
+            ORDER BY seq
+                LIMIT 0, #{size}
```
이전에 피드백 주신 `WHERE IN` 파라미터 제한이 옵션값에 설정된 packet 크기만큼만 가능한 것으로 확인했습니다. 

아직 in 절의 파라미터가 packet 용량을 넘지는 않지만,
WHERE IN 쿼리 결과 사이즈가 너무 커서 페이징 처리로 변경했습니다. 
일반 offset에서는 인덱스가 뒤로 갈수록 검색 효율이 떨어져서 
no offset 쿼리로 작성했습니다. 

> The number of values in the IN() list is only limited by the [max_allowed_packet](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) value.
([링크](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in))

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/AggregateSubscriber.java:
```diff
@@ -0,0 +1,34 @@
+            map.put(key, longAdder);
+            log.info("consumer {}", map);
+        });
+    }
```
`LongAddr` 클래스가 연산의 원자성을 보장해주기 때문에 
`ConcurrentHashMap`을 사용하지 않았습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -25,46 +28,59 @@
     private final TicketOrderAggregationMapper aggregationMapper;
+    private final AggregateSubscriber subscriber;
+    private final ExecutorService es = Executors.newFixedThreadPool(3);
+    private final static int SIZE = 2_000_000;
```
크기별로 테스트 하느라 크게 잡았습니다!

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -25,46 +28,59 @@
+                .map(entry -> new Response(
+                        concertSeq, entry.getKey(), groupBy.get(entry.getKey()).size(), entry.getValue().intValue())
+                ).toList();
     }
```
`es.execute()` 내부에서 페이징 처리를 하고 있는데
페이징이 한 스레드에서 처리되어 멀티 스레드 이점을 못살리고 있습니다.

gutenLEE replied:
집계 처리에 
멀티스레드 : 9초
스트림 : 14초
싱글 스레드 11초
소요 되었습니다. 

페이징 별로 스레드 할당하는 부분을 개선하면 처리 시간 향상이 있을 것을 예상합니다.

gutenLEE committed:
feat : 멀티스레딩 1차 개발 완료

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/MultiAggregationReader.java:
```diff
@@ -25,7 +26,10 @@ public class MultiAggregationReader implements AggregationReader {
-        return strategy.execute(request.concertSeq(), tickets);
+        strategy.execute(request.concertSeq(), tickets);
+
+        return Collections.emptyList();
```
비동기 멀티스레딩 처리에서 연산 결과를 취합까지 대기하는 로직이 
Thread Per Request 모델에서는 적합하지 않다 생각하여 
`CountDownLatch` 로 스레드 대기 로직 제거했습니다. 
그래서  empty 리스트 반환하고 있습니다. 
리턴 타입은 결과를 어떻게 보여줄지 설계가 되면 통일작업 하겠습니다.

f-lab-maverick replied:
작업이 완료된 후 집계된 결과를 리턴하는게 목적이라면 리턴값을 받은 후 내려주는게 적절해보입니다.

DB에 집계결과가 저장되면 된다면 이상태로 비동기 리턴처리를 해두어도 되겠네요. 물론 그 경우 리턴값은 명시적으로 void와 같이 변경되어야 할 것입니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/firstclass/GradeInfo.java:
```diff
@@ -0,0 +1,62 @@
+        this.groupBy = initMinMax(map);
+    }
+
+    public String getGrade(long ticketSeq){
+        return groupBy.entrySet().stream()
+                .filter(entry -> {
+                    MinMax minMax = entry.getValue();
+                    return minMax.isWithinRange(ticketSeq);
+                })
+                .findFirst()
+                .map(Map.Entry::getKey)
+                .orElseThrow(() -> new IllegalArgumentException("티켓 시퀀스에 속하지 않는 시퀀스입니다 : " + ticketSeq));
+    }
```
이전에는 티켓 등급별로 쿼리로 조회 해왔는데 
랜덤하게 조회하고 어플리케이션에서  등급별로 추리기 위해 추가된 기능입니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/firstclass/ResultMap.java:
```diff
@@ -0,0 +1,44 @@
+                .toList();
+    }
+
+    private LongAdder getInit() {
+        LongAdder longAdder = new LongAdder();
+        longAdder.increment();
+        return longAdder;
+    }
```
LongAddr 클래스에는 `addAndGet` 같은 메서드가 없어 이렇게 리턴합니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -28,59 +29,28 @@
-            latch.await();
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
+        for (long offset : offsets) {
+            String uuid = publisher.generateUuid();
+            CompletableFuture.supplyAsync(() -> {
+                return aggregationMapper.selectByTicketSeqBetween(new ZeroOffsetRequest(concertSeq, start, to, offset, SIZE));
+            }, es).thenAcceptAsync(res -> publisher.publish(new TaskMessage<>(uuid, concertSeq, offsets.size(), res, gradeInfo)), es);
```
이전 로직에서 페이징 SIZE를 줄여서 청크가 많을수록 조회 시간이 오래 걸렸습니다. 
원인 분석 결과 멀티스레드로 작업을 할당해도 
I/O 작업에서 blocking으로 스레드가 대기 상태에 빠져서 멀티스레드 효과가 크지 못했습니다. 

비동기 처리하여 스레드가 I/O 작업으로 인한 대기 상태에 빠지지 않게 했습니다.

f-lab-maverick replied:
현재 로직은 특정 콘서트에 발급된 티켓의 가장 최소 seq, 최대 seq를 구한 다음 균등하게 분할하여 처리하는 로직이네요.
만약 여러 콘서트가 동시에 티케팅이 되고있어서 서로 다른 콘서트의 seq가 섞인 상태라면 꽤나 균등하지 못한 처리가 될 가능성이 있지 않을까요?

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/subscriber/AggregateSubscriber.java:
```diff
@@ -0,0 +1,61 @@
+
+    @ThreadSafe
+    private final ResultMap result = new ResultMap();
+    private final Set<String> uuidSet = new HashSet<>();
```
비동기 처리에서 작업 완료 여부를 확인 하는 부분이 어려웠네요..
작업별로 `UUID` 발급해서 `set` 자료구조에 추가한 뒤 `set.size()` 와  `taskMassage`에 담아준 
카운트 일치 비교로 완료 여부 확인하게 했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/subscriber/AggregateSubscriber.java:
```diff
@@ -0,0 +1,61 @@
+
+        collect.entrySet().forEach(result::add);
+
+        if(doneTask(taskMessage.totalCount())) {
+            resultDBTransfer.transferAll(result.toList(taskMessage));
+        }
```
결과 DB 저장합니다.

gutenLEE commented on src/main/resources/mapper/TicketOrderAggregationtMapper.xml:
```diff
@@ -78,13 +76,20 @@
                 seq > #{offset}
-            ORDER BY seq
-                LIMIT 0, #{size}
+            LIMIT 0, #{size}
```
no offset 쿼리를 사용하려 했는데 현재 테이블의 컬럼으로는 
no offset 사용이 불가했습니다.. 

하여 쿼리 성능 향상을 위해 no offset 쿼리를 선택하고 
등급별로 추리는 작업은 어플리케이션에서 하는 것으로 변경했습니다.

gutenLEE commented on src/main/resources/mapper/TicketOrderAggregationtMapper.xml:
```diff
@@ -78,13 +76,20 @@
+    <!--티켓 등급별 예매 시도 조회-->
+    <select id="selectMinMax" parameterType="com.heeverse.ticket_order.domain.dto.persistence.AggregateSelectMapperDto$ZeroOffsetRequest"
+            resultType="com.heeverse.ticket_order.domain.dto.persistence.AggregateSelectMapperDto$MinMaxResponse">
+        SELECT
+            min(seq) minSeq,
+            max(seq) maxSeq
+        FROM ticket_order_log_copy t
+        WHERE
+            concert_seq = #{concertSeq}
+        ORDER BY seq
```
count 함수는 데이터 전체를 스캔하며 집계하기 때문에 응답까지 4초가 소요되었습니다. 
이 쿼리는 응답까지 0.5초 아래로 떨어졌는데
concert_seq 인덱스로 범위 지정 후 min, max 함수에서
첫 행과 마지막 행만을 읽어서 빨라진 게 아닌가 생각됩니다.

gutenLEE reviewed (commented):
![집계-pub-sub](https://github.com/f-lab-edu/heeverse-ticket/assets/61870748/3a3b5436-b20a-47ff-86ad-e5306197dcca)

Pub-Sub 구조로 멀티스레딩 처리 했습니다. 
이전 PR에서는 chuck가 많아질수록 처리 시간이 길어져 
bulk 로 데이터를 한 번에 받아와야 9초가 걸렸습니다. 

변경된 코드에서는 chuck로 멀티스레딩 처리해도 7~8.5 초로 개선했습니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/domain/dto/persistence/AggregateInsertMapperDto.java:
```diff
@@ -15,10 +15,15 @@ public class AggregateInsertMapperDto {
+        this.gradeName = gradeName;
+        this.totalTickets = totalTickets;
+        this.orderTry = (int) orderTry;
+    }
```
long 타입을 int로 형변환을 하고 있네요. overflow와 관련된 이슈는 없을까요?

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -0,0 +1,56 @@
+    public void execute(long concertSeq, List<Ticket> tickets) {
+
+        GradeInfo gradeInfo = new GradeInfo(tickets);
+        long start = TicketUtils.minSeq(tickets);
+        long to = TicketUtils.maxSeq(tickets);
```
이렇게 min, max를 따로 구하지 않고, 한번만 정렬한다면 min, max를 한번에 가져올 수 있지 않을까요?
상당히 많은 데이터일텐데 정렬은 꽤 리소스가 많이 들어가는 작업입니다. 정렬 횟수를 줄일 수 있다면 최대한 줄여주세요.

f-lab-maverick replied:
이미 만들어두신 `MinMaxResponse`로 리턴하게 만들어도 될거같네요.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/firstclass/ResultMap.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public void add(Map.Entry<String, Long> entry) {
+        resultMap.computeIfAbsent(entry.getKey(), r -> getInit());
+        resultMap.computeIfPresent(entry.getKey(), (key, a) -> {
+            a.add(entry.getValue());
+            return a;
+        });
```
`compute`라는 메서드로 단일연산으로 만들어주는게 더 좋아보입니다.

`computeIfAbsent` 후 `computeIfPresent`를 호출하기 때문에 동시성 이슈가 발생할일은 거의 없어보이는데요, 원칙적으로 두가지 이상의 연산을 락 없이 진행할 때에는 동시성 이슈가 발생할 수 있습니다.

concurrent hash map이 동시성을 제어하는 방법은 두가지 이상의 연산을 하나로 묶을 수 있는 기능을 제공하기 때문입니다. 그런데 이런 연산을 여러 스레드레서 각각 여러번 호출하면 당연하게도 동시성 이슈가 발생할 수 있습니다. 그러므로 아예 하나의 원자적인 연산을 만들어서 처리해주면 동시성 관련 고민을 안해도 될 것으로 보이네요~
```suggestion
        resultMap.compute(entry.getKey(), r -> r == null ? getInit() : r.add(entry.getValue());
```

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/subscriber/AggregateSubscriber.java:
```diff
@@ -0,0 +1,61 @@
+@Slf4j
+public class AggregateSubscriber implements Subscriber<List<SimpleResponse>> {
+
+    @ThreadSafe
+    private final ResultMap result = new ResultMap();
```
이 resultMap은 Spring 전역에서 사용할텐데요. 집계가 시작하고 끝날때마다 모두 초기화해주는 로직이 필요해보입니다.
또한, 집계가 동시에 여러건이 돌 경우 당연하겠지만 이 result 객체를 같이 사용하게되므로 고려해주셔야 합니다.

이런 점을 막으려면 이 `AggregateSubscriber.result`객체를 공용으로 Bean에 넣어두는게 아닌 작업하는 스레드끼리만 공유할 수 있도록 파라미터로 넣어주거나, 다른 방식을 고려해야겠네요.

f-lab-maverick reviewed (changes requested):
꼼꼼하게 확인해보느라 조금 늦었습니다.

점점 완성도가 높아지는것이 보이고 있습니다. 리뷰 확인 후 수정 부탁드려요~

gutenLEE committed:
feat : 멀티스레딩 처리 PR 피드백 반영

gutenLEE committed:
feat : 스트림 & 싱글스레드 버전 완료

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/AsynchronizableAggregation.java:
쿼리로 집계 버전과 프로그래밍 집계 버전 리턴 타입이 달라서 
인터페이스를 `AsynchronizableAggregation`, `SynchronizableAggregation` 두 개로 분리했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/SingleThreadStrategy.java:
싱글스레드도 멀티스레드와 스트림처럼 동일한 chunk 단위로 수행해야
테스트 결과 의미가 있을듯하여 동일한 작업 단위 처리로 변경했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/StreamAggregationStrategy.java:
```diff
@@ -0,0 +1,52 @@
+        List<List<Long>> chunks = PaginationProvider.toChunk(collectedTicketSeq, CHUNK_SIZE);
+        log.info("chunk size {}", chunks.size());
+
+        ResultMap resultMap = new ResultMap();
+        Collections.synchronizedList(chunks)
```
앞서 피드백주신 스트림 병렬 처리시에도 
동시성 이슈를 방지하기 위해 스레드 세이프한 자료구조 사용으로 변경했습니다.

f-lab-maverick replied:
`synchronizedList`과 `CopyOnWriteArrayList`를 비교해보는 작업을 해보는건 어떠실까요? `CopyOnWriteArrayList`도 고려 대상에 포함되었을까요?

gutenLEE replied:
아하 비교해보겠습니다!

gutenLEE replied:
비교 결과  읽기 작업만 수행할 경우 `CopyOnWriteArrayList` 이 더 적합했습니다. 
`CopyOnWriteArrayList`은 병렬성을 위해 불변 객체를 외부에 공개하여 순회 시 동기화가 필요 없습니다. 

그러나 Iterator()로 복사된 객체를 얻을 경우 스냅숏처럼
호출 시에 복사된 객체를 얻기 때문에 그 이후에 원본 리스트에 변경된 내용은
조회할 수 없습니다. 

현재 구조에서는 원본 리스트에 변경이 발생하지 않음, 리스트 사이즈 크지 않음, 콜렉션 조회만 필요
등의 이유로 `CopyOnWriteArrayList` 을 사용하는게 적합할 것 같습니다. 

관련 내용은 블로그에 정리해두었습니다. 
https://guui-dev-lee.tistory.com/17

f-lab-maverick replied:
핵심을 잘 비교해주셨네요~ 고생하셨습니다.

gutenLEE commented on src/main/java/com/heeverse/common/util/PrimitiveUtils.java:
```diff
@@ -0,0 +1,18 @@
+@UtilityClass
+public class PrimitiveUtils {
+
+    public static int toIntSafely(long value) {
+        if (value < Integer.MIN_VALUE || value > Integer.MAX_VALUE) {
+            throw new ArithmeticException("overflow 발생합니다 : " + value);
+        }
+        return (int) value;
+    }
```
타입을 모두 long으로 통일할까 하다가 
형변환이 필요한 부분만 안전하게 int 타입으로 형변환하도록 했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/TicketAggrFacade.java:
읽기 전용 트랜잭션을 적용하기 위해 파사드 패턴 사용했습니다. 
격리 수준은 READ_COMMITED로 낮췄습니다.

f-lab-maverick replied:
격리수준을 낮춤으로써 발생하는 문제에 대해서는 어떤 고민을 하셨나요?

gutenLEE replied:
조회 대상 테이블은 `INSERT` 작업만 발생하고 `UPDATE`는 발생하지 않습니다. 
그리고 트랜잭션마다 1건의 `SELECT` 쿼리만 존재합니다. 

따라서 `READ_COMMITED` 격리 수준에서 발생할 수 있는 
한 트랜잭션에서 여러번 조회 했을 때 보였다 안보였다 하는 `Phantom read` 문제가 발생하지 않기 때문에 
집계를 위한 조회에서는 격리 수준을 낮춰도 괜찮다 판단했습니다.

gutenLEE commented on src/main/resources/mapper/TicketOrderAggregationtMapper.xml:
```diff
@@ -55,7 +55,7 @@
 
     <!--티켓 등급별 예매 시도 조회-->
-    <select id="selectByTicketSeqList" parameterType="java.lang.Long"
+    <select id="selectTicketSeqWhereIn" parameterType="java.lang.Long"
```
피드백 주신 여러 콘서트 예매가 동시에 진행될 경우 seq의 순차성을 보장할 수 없어 
다른 페이징 전략을 고민해봤습니다. 

상태 공유하지 않는 페이징 처리를 고민한 결과 세 가지 방법이 나왔습니다. 
1. row_number()로 행 번호를 부여하여 no offset 페이징 하기
2. Count  함수로 전체 개수를 구하여 offset 방식의 페이징
3. WHERE IN 절에 파라이터 개수 분할해서 처리하기

모두 트레이드오프가 존재했으나 현재 구조에서 가장 성능이 좋은 3번을 선택했습니다.

f-lab-maverick replied:
IN절도 많아질 경우 성능상 안좋을 수 있습니다. 실행계획을 확인해서 적용해주세요.

gutenLEE replied:
> IN절도 많아질 경우 성능상 안좋을 수 있습니다. 실행계획을 확인해서 적용해주세요.

![조회 시간 비교](https://github.com/f-lab-edu/heeverse-ticket/assets/61870748/ab140ac4-a9b2-48b8-9227-7bed046964fc)

이전에 하나의 페이지로만 성능 비교를 수행했는데, 
전체 데이터 조회로 성능 비교를 해보니 No Offset 방식이 가장 빨랐습니다. 

관련 내용은 블로그에 기재해두었습니다. 
https://guui-dev-lee.tistory.com/16

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -28,28 +32,28 @@
+        log.info("chunk size {}", chunk.size());
 
-        for (long offset : offsets) {
+        AggregateSubscriber subscriber = new AggregateSubscriber();
```
`AggregateSubscriber` 을 `Bean`으로 사용하면 상태를 갖게되어 초기화 메서드가 필요했는데
안전하게 필요할 때마다 생성하는 것으로 변경했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -28,28 +32,28 @@
-            CompletableFuture.supplyAsync(() -> {
-                return aggregationMapper.selectByTicketSeqBetween(new ZeroOffsetRequest(concertSeq, start, to, offset, SIZE));
-            }, es).thenAcceptAsync(res -> publisher.publish(new TaskMessage<>(uuid, concertSeq, offsets.size(), res, gradeInfo)), es);
+            CompletableFuture.supplyAsync(() -> ticketAggrFacade.read(paging), ioBoundWorker)
+                    .thenAcceptAsync(res -> publisher.publish(
+                            new TaskMessage<>(uuid, concertSeq, chunk.size(), res, gradeInfo), subscriber), cpuBoundWorker);
```
앞선 PR에서는 하나의 스레드풀에서 I/O, CPU 작업을 함께 할당했는데
각각 스레드풀에서 실행하도록 변경했습니다.

f-lab-maverick replied:
코드가 이제 한눈에 읽히네요 👍

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/domain/dto/AggregateDto.java:
```diff
@@ -15,31 +18,47 @@ public static class Request {
+            return strategyType == StrategyType.QUERY;
         }
     }
 
 
     @Getter
     public static class Response {
 
-        private final Long concertSeq;
-        private final String gradeName;
-        private final Integer totalTickets;
-        private final Integer orderTry;
+        private Long concertSeq;
+        private String gradeName;
+        private Integer totalTickets;
+        private Integer orderTry;
+        private String message;
```
nullable이라 final 키워드를 제거한 것이라면, 해당 값이 null이 들어갈 수 있다는 의미인 `@Nullable`을 붙여주면 사용할 때에도 IntelliJ에서도 해당 값이 Null일 수 있다며 경고를 표시해줍니다~

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/StreamAggregationStrategy.java:
```diff
@@ -0,0 +1,52 @@
+        Collections.synchronizedList(chunks)
+                .parallelStream()
+                .map(ticketAggrFacade::read)
+                .flatMap(list -> {
+                    Map<String, Long> map = list.stream().collect(Collectors.groupingBy(res -> gradeInfo.getGrade(res.ticketSeq()), Collectors.counting()));
+                    return map.entrySet().stream();
+                })
```
```suggestion
                .flatMap(list -> list.stream().collect(Collectors.groupingBy(res -> gradeInfo.getGrade(res.ticketSeq()), Collectors.counting())).entrySet().stream())
```

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/StreamAggregationStrategy.java:
```diff
@@ -0,0 +1,52 @@
+                    Map<String, Long> map = list.stream().collect(Collectors.groupingBy(res -> gradeInfo.getGrade(res.ticketSeq()), Collectors.counting()));
+                    return map.entrySet().stream();
+                })
+                .forEach(resultMap::add);
+
+        resultDBTransfer.transferAll(resultMap.toList(concertSeq));
```
병렬스트림으로 돌리면 resultMap에 모든 작업이 끝나기 전 `toList`를 호출할 가능성은 없을까요?

gutenLEE replied:
학습 테스트를 작성해본 결과
병렬 스트림에서 terminal operator인 `forEach()` 는 스트림 요소의 consume 만을 보장하고
consume  내부의 모든 로직의 완료는 보장하지 않았습니다. 

문제 상황 구현 코드 로직
```
1. 100개의 요소를 consume하면 소비했음을 체크
2. 스레드 슬립 랜덤하게 걸기
3. 스레드 슬립이 종료되면 `list`에 소비한 요소 추가
```
예상한 결과는 `list.size() == 100` 이였으나 실제로는 랜덤하게 size가 변경되는 것을 확인했습니다. 

스트림 파이프라인 내부에서 `collect()` 같은 오퍼레이터를 사용해서 취합하도록 검토해야겠습니다.

```java
    @Test
    void consumeTest() throws Exception {
        ArrayList<Object> list = new ArrayList<>();
        int[] check = new int[100];

        IntStream.range(0, 100)
                .parallel()
                .forEach(i -> {
                    check[i] += 1;
                    threadSleepRandomly();
                    list.add(i);
                });

        // 0 ~ 100 요소 consume 확인
        Arrays.stream(check).forEach(i -> Assertions.assertEquals(1, i));
        
        // consume의 모든 코드 완료 확인
        Assertions.assertEquals(100, list.size(), "list 사이즈는 100 이다");
    }
```

f-lab-maverick reviewed (commented):
스트림 처리가 빠른것은 병렬스트림으로 돌려서 사실상 병렬처리를 했기 때문으로 보입니다.
그렇다면 병렬로 직접 만든 로직쪽에도 개선이 될 여지가 있지 않을까요? 성능상 큰 차이는 없었지만, 코드를 어느정도 정리하고, 로직을 다듬으면 머지해도 좋을 것으로 보입니다. 조금만 더 다듬어주세요~

gutenLEE committed:
chore : 주석 추가

gutenLEE committed:
refactor : 결과 취합 Map 제네릭&상속 변경

gutenLEE committed:
refactor : reader - strategy 구조 변경

1. 3개의 reader 중복 로직을 CommonAggregationReader 로 취합하고 구현체 삭제
2. Strategy 로 전달하는 파라미터 Wrapping 객체로 전달
3. Publisher 제거

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/firstclass/ResultMap.java:
`Map` 을 반복적으로 사용하는 패턴이 보여 일급컬렉션 클래스를 제네릭&추상클래스로 추상화했습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/producer/Publisher.java:
아직 Pub-Sub 패턴을 사용해야 할 명확한 로직이 없어 제거했습니다.

ZhenxiKim replied:
MultithreadingStrategy에서 데이터를 읽어오고 집계를 하지만, 멀티스레드 pub-sub구조를 사용하여 publisher쪽에서는 데이터를 읽기만하는 쓰레드, subscriber에는 읽어온 데이터를 집계하는 패턴도 적용해 볼 수 있을것 같습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/AggregationJobWrapper.java:
전략 객체 파라미터 개수가 많아 `wrapper` 클래스로 래핑해서 전달로 변경했습니다

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/MultithreadingStrategy.java:
```diff
@@ -30,28 +28,31 @@
             CompletableFuture.supplyAsync(() -> ticketAggrFacade.read(paging), ioBoundWorker)
-                    .thenAcceptAsync(res -> publisher.publish(
-                            new TaskMessage<>(uuid, concertSeq, chunk.size(), res, gradeInfo), subscriber), cpuBoundWorker);
+                    .thenApplyAsync(res -> subscriber.subscribe(new TaskMessage<>(concertSeq, res, gradeInfo)), cpuBoundWorker)
+                    .thenAcceptAsync(resultMap -> saveIfDone(concertSeq, taskCount, resultMap), cpuBoundWorker);
+        }
```
작업 결과 취합과 DB 저장을 `Subscriber` 객체에서 전략객체로 옮겼습니다.

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/StreamAggregationStrategy.java:
```diff
@@ -27,26 +31,35 @@ public class StreamAggregationStrategy implements AggregationStrategy {
+        Map<String, Long> result = getTicketOrderLog(chunks)
+                .map(list -> getGroupByGrade(gradeInfo, list))
+                .flatMap(StreamHelper::toEntrySetStream)
+                .collect(Collectors.groupingByConcurrent(
+                        Map.Entry::getKey,
+                        Collectors.summingLong(Map.Entry::getValue)
+                ));
```
결과 취합을 스트림 terminal operator에서 하도록 변경했습니다. 
병렬 스트림을 사용하여 `groupingByConcurrent` 사용했습니다.

ZhenxiKim replied:
stream 계산을 취합하고 넣는 이유가 있나요? 중간에 완성된 결과를 디비에 넣는것이 안전하지않을까요?

gutenLEE replied:
대면으로 피드백했습니다!

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/CommonAggregationReader.java:
기존 3개의 `Reader` 구현체에 중복되는 로직이 보여 통일했습니다.

gutenLEE reviewed (commented):
<img src="https://github.com/f-lab-edu/heeverse-ticket/assets/61870748/550b86b3-417d-4a2e-86c3-5080e4caaafb" style="witth: 500px;">

전반적으로 리팩토링 작업을 수행했습니다.  
중복된 패턴들이 보여 구조 개선했습니다.

[변경 전] 

- 3개의 Reader 구현체가 각각의 전략 객체를 의존  

[변경 후]

- 중복 로직을 제거하고 CommonAggregationReader의 메서드 호출 시 전략 객체를 파라미터로 전달

gutenLEE committed:
refactor : 객체 참조 변수명 변경

gutenLEE commented on src/main/java/com/heeverse/ticket_order/service/reader/subscriber/AggregateSubscriber.java:
```diff
@@ -18,7 +18,7 @@
 
     @ThreadSafe
-    private final ResultConcurrentMap result = new ResultConcurrentMap(new ConcurrentHashMap<>());
+    private final ResultConcurrentMap resultConcurrentMap = new ResultConcurrentMap(new ConcurrentHashMap<>());
```
개구리책 112 페이지를 보다가 **객체 참조 변수명은 객체스럽게 정하라**는 의견을 반영하였습니다. 
`ConcurrentMap`을 멤버로 갖는 `resultMap` 객체여서 `resultConcurrentMap` 으로 변경했습니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/firstclass/ResultConcurrentMap.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class ResultConcurrentMap extends ResultMap<String, LongAdder, Long> {
+
+    public ResultConcurrentMap(Map<String, LongAdder> map) {
+        super(map);
```
`ConcurrentHashMap`을 생성자로 받고 싶은 것이라면 `Map` 인터페이스로 생성 파라미터를 받지 말고 `ConcurrentMap` 인터페이스를 사용해주세요. `Map`타입으로 받으면 HashMap 등 concurrent와는 관계없는 객체가 들어올 수 있지 않을까요?

gutenLEE replied:
아 확장한 클래스니까 추상화된 타입으로 받을 필요가 없었네요... 수정하겠습니다.

f-lab-maverick commented on src/main/java/com/heeverse/ticket_order/service/reader/strategy/SingleThreadStrategy.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    public void execute(AggregationJobWrapper jobWrapper) {
+
+        ResultHashMap resultHashMap = new ResultHashMap(new HashMap<>());
```
이렇게 생성자로 `new HashMap`을 생성해주는것도 관습적인 중복 코드가 될 수 있겠네요.
`ResultHashMap`의 기본 생성자로 생성시 내부적으로 `new HashMap<>()`를 사용하도록 해주는건 어떠신가요?

gutenLEE replied:
넵!

f-lab-maverick commented on src/test/java/com/heeverse/ticket_order/service/reader/PaginationProviderTest.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+
+        @Test
+        void syncListTest() throws Exception {
+
+
+
+        }
+
```
사용하지 않는 테스트는 제거해주세요

f-lab-maverick commented on src/test/java/com/heeverse/ticket_order/service/reader/PaginationProviderTest.java:
```diff
@@ -0,0 +1,224 @@
+
+        @Test
+        @DisplayName("WHERE IN 쿼리")
+        @Order(3)
```
테스트끼리는 서로 영향을 미치면 안됩니다. `@Order(3)` 를 작성하신 이유가 어떤것인가요?

gutenLEE replied:
아 로컬에서 테스트할 때 빨리 끝나는 테스트부터 순서대로 돌리려고 임시로 추가했는데 
제거를 깜빡했네요. 수정하겠습니다

f-lab-maverick reviewed (approved):
고생하셨습니다~ 머지하셔도 좋습니다. 코멘트만 한번 확인해주세요.

gutenLEE committed:
refactor : resultMap 리팩토링
